### DISCUSSION TITLE
What's the best way to update Quill content from deltas on the server?

### DISCUSSION THREAD

**Discussion by owenfar** [slab/quill#4026] [Q&A]
I have recently implemented Quill in my project and it's great! Very easy to understand and documentation is well written. But I can't seem to find any best practices on how to update content (delta object) from `Delta` (coming from `text-change` event) on the server-side.

When one is not using Node.js on the backend, how can we update the changes coming from deltas (`text-change` event) on the server when using other languages such as Python or PHP? While deltas are great, it's hard to quickly implement a parser that can do the same as `setContents` that is available in the library. The `retain` does not work with `getContents`, `getText` or `getSemanticHTML`. To apply incremental changes, I would need to re-write the whole logic there. But that would be overkill, I'm probably missing something...

I really wish to avoid sending large content for every simple update, it doesn't make sense.

Can someone help? Did anyone else find this or solved this issue somehow?

Your time is very much appreciated!

**Comment by benbro** [comment]
I suggest to start here:
https://quilljs.com/docs/api#text-change
https://quilljs.com/docs/delta
https://quilljs.com/guides/designing-the-delta-format

**Reply by owenfar** [reply]
Hello @benbro, thank you for your suggestions. But sorry if I wasn't clear enough.

I already read everything that can be read on the website, I also searched around the web and went through quill-awesome, but the question still remains. 

How are we suppose to use the information we get from delta to update the data on backend services that can't run JS?

As far as I can tell, the delta only works best with Quill library, but I can't for example easily use it to submit the same changes on the backend without deeply understanding how it works and carefully implementing a parser on the backend that updates the content. Of course the easy way would be to keep sending the whole content for every little update, but that was the point, that it's hard to believe that everyone handles it that way.

Whichever type I get, whether it's getContents, getText, or getSemanticHTML, I cannot easily use the delta without carefully building my own parser on the backend (e.g. Python, PHP, etc.).

I understand that it's a front-end library, but there must have been some thought around handling delta requests to the server.

**Reply by owenfar** [reply]
Hello @benbro, should I open an issue to ask for questions? Where's the best place to get help? I asked the same on StackOverflow but I got a downvote.. I don't know what's wrong with my question 😄

**Comment by benbro** [comment]
There are third party Delta libs here:
https://github.com/quilljs/awesome-quill?tab=readme-ov-file#deltas

**Reply by owenfar** [reply]
Thank you for your response. I opened another discussion in one related library, because I had already checked these and none really seem to do what I'm looking for. But I really appreciate your time for helping me. I will also update again the question here as I'm still not being very clear with what I wish to achieve.

**Comment by benbro** [comment]
If you have one user (not concurrent editing) you need to compose the delta on the server with the changes from the client.
https://github.com/quilljs/delta?tab=readme-ov-file#compose
If you can't find a delta lib in the language you use on the server you'll need to translate it from the quilljs/delta (Typescript) or similar.
